crates/runtime/src/spice_metrics_recorder.rs

```diff
@@ -0,0 +1,231 @@
+    UnableToParsePrometheusMetrics { source: std::io::Error },
+}
+
+pub struct SpiceMetricsRecorder {
+    socket_addr: Arc<SocketAddr>,
+    table: Arc<MetricsTable>,
+}
+
+impl SpiceMetricsRecorder {
+    pub fn table(&self) -> Arc<MetricsTable> {
+        Arc::clone(&self.table)
+    }
+}
+
+impl SpiceMetricsRecorder {
+    pub fn new(socket_addr: SocketAddr) -> Self {
+        let table = MetricsTable::new();
+
+        Self {
+            socket_addr: Arc::new(socket_addr),
+            table: Arc::new(table),
+        }
+    }
+
+    async fn tick(socket_addr: &SocketAddr, table: &Arc<MetricsTable>) -> Result<(), Error> {
+        let body = reqwest::get(format!("http://{socket_addr}/metrics"))
+            .await
+            .context(FailedToQueryPrometheusMetricsSnafu)?
+            .text()
+            .await
+            .context(FailedToQueryPrometheusMetricsSnafu)?;
+
+        let lines = body.lines().map(|s| Ok(s.to_owned()));
+        let scrape =
+            prometheus_parse::Scrape::parse(lines).context(UnableToParsePrometheusMetricsSnafu)?;
+
+        for sample in scrape.samples {
+            let value: f64 = match sample.value {
+                prometheus_parse::Value::Counter(v)
+                | prometheus_parse::Value::Gauge(v)
+                | prometheus_parse::Value::Untyped(v) => v,
+                prometheus_parse::Value::Histogram(v) => v.into_iter().map(|v| v.count).sum(),
+                prometheus_parse::Value::Summary(v) => v.into_iter().map(|v| v.count).sum(),
+            };
+
+            table
+                .add_record(
+                    sample.timestamp.timestamp(),
+                    &sample.metric,
+                    value,
+                    sample.labels.to_string().as_str(),
+                )
+                .await?;
+        }
+
+        Ok(())
+    }
+
+    pub fn start(&self) {
+        let table = Arc::clone(&self.table);
+        let addr = Arc::clone(&self.socket_addr);
+
+        spawn(async move {
+            loop {
+                if let Err(err) = SpiceMetricsRecorder::tick(&addr, &table).await {
+                    tracing::debug!("{err}");
+                }
+                tokio::time::sleep(Duration::from_secs(30)).await;
+            }
+        });
+    }
+}
+
+pub type PartitionData = Arc<RwLock<Vec<RecordBatch>>>;
+
+pub struct MetricsTable {
+    pub schema: SchemaRef,
+    pub batches: RwLock<Vec<PartitionData>>,
+    constraints: Constraints,
+}
+
+impl Default for MetricsTable {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl MetricsTable {
+    pub fn new() -> Self {
+        let fields = vec![
+            Field::new(
+                "timestamp",
+                DataType::Timestamp(TimeUnit::Second, None),
+                false,
+            ),
+            Field::new("metric", DataType::Utf8, false),
+            Field::new("value", DataType::Float64, false),
+            Field::new("labels", DataType::Utf8, false),
```

Comment:
I think it could be good to have labels as a DataType::Map or something where its more SQL-queryable. I don't know if its possible since i think the values in Map/Dict type must be consistent. alternatively have UTF8 as JSON.

Reply:
Arrow doesn't support non strict types, so I'll leave `labels` as string for now. I'm planning to extend it with `tags` field, which will hold some known tags like `dataset` and `engine`.